# Massachusetts Colleges Capstone Project - Raul Gonzalez
# Implementation
# April 7th, 2020

Choosing which college or university to attend is one of the biggest decisions that you will make in terms of your personal and professional development, and it can be overwhelming if you don’t know where to start. College is important for many reasons, including long-term financial gain, job stability, career satisfaction and success outside of the workplace. In essence, college teaches us how to learn and grow in every aspect of our lives. As Ralph Waldo Emerson says, "The things taught in schools and colleges are not an education, but the means to an education." Of course, the quality of the education offered is a very important factor in deciding which college to attend, as well as the overall on-campus experience. However, another important factor is the off-campus experience offered, including the historical, cultural and social background of the college's city since at the end attending college is preparing us to become active members of our societies.

Most of the time prospective college students can't visit the city of the college they are planning to attend because of time or economic limitations and are sometimes insecure if they are going to like the environment offered by that city. Also, many students choose to attend a certain college just because they really like the city in which it is located. It seems that knowing in advance the different types of environments offered by the cities would facilitate prospective college students the important task of choosing which college or university to attend. Maybe a student would prefer to attend a college in a city with a lot of Italian restaurants in the surroundings or maybe a city with many museums and parks to visit during weekends.

With its incredibly significant, undeniably unique place in American history and culture, from the earliest historical period of colonial America onward, Massachusetts continues to play a primary contributing roll to American high-culture and fine-arts. Massachusetts is home to countless world-class museums and national historical sites and has produced some of Americans most famously creative academics, artists, writers, and musicians. Massachusetts’ role in American education is also without equal. Massachusetts is home to the United States’ oldest high school, the first public library, oldest boarding school, oldest college, and the first women’s college. Additionally, top level universities such as Harvard and MIT, which consistently rank among the world’s best universities year after year, are located in this state. Massachusetts has 12% of the top research universities and 15% of the top 40 liberal arts colleges. Several of the world’s best medical and technology facilities are located here as well as numerous multinational corporations. In summary, the state of Massachusetts seems like a great place to decide to accomplish college studies. This project aims to cluster of find the different environments offered by the cities of Massachusetts that houses the state´s colleges and universities, seeking to facilitate prospective college students, wanting to study there, the important task of choosing the right college.

### Extracting data from Wikipedia using pandas

In [38]:
# extract tables from wikipedia
from pandas.io.html import read_html

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import re

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')
page = 'https://en.wikipedia.org/wiki/List_of_colleges_and_universities_in_Massachusetts'

wikitables = read_html(page,  attrs={"class":"wikitable sortable"})

print ("Extracted {num} wikitables".format(num=len(wikitables)))

Libraries imported.
Extracted 3 wikitables


We will only need the first table for the implementation

In [39]:
df=wikitables[0]
print(df.shape)
df.head(10)

(104, 7)


School  Location[note 1]  \
0             American International College       Springfield   
1                            Amherst College           Amherst   
2                         Anna Maria College            Paxton   
3                         Assumption College         Worcester   
4                             Babson College         Wellesley   
5               Bard College at Simon's Rock  Great Barrington   
6                        Bay Path University        Longmeadow   
7                          Bay State College            Boston   
8                             Becker College         Worcester   
9  Benjamin Franklin Institute of Technology            Boston   

               Control[1]                            Type[1] Enrollment[16]  \
0  Private not-for-profit                Master's university      2,177[17]   
1  Private not-for-profit              Baccalaureate college      1,817[18]   
2  Private not-for-profit                Master's university      1,455[19]   
3  Private not-for-profit                Master's university      2,813[20]   
4  Private not-for-profit          Special-focus institution      3,250[21]   
5  Private not-for-profit  Baccalaureate/associate's college        354[22]   
6  Private not-for-profit              Baccalaureate college      2,370[23]   
7              For-profit                Associate's college      1,721[24]   
8  Private not-for-profit              Baccalaureate college      1,826[25]   
9  Private not-for-profit          Special-focus institution        475[26]   

    Founded              Accreditation[16]  
0  1885[17]    AOTA, APTA, CCNE, NEASC[17]  
1  1821[18]                      NEASC[18]  
2  1946[19]         NASM, NEASC, NLNAC[19]  
3  1904[20]                      NEASC[20]  
4  1919[21]                      NEASC[21]  
5  1964[22]                      NEASC[22]  
6  1897[23]                AOTA, NEASC[23]  
7  1946[24]  ABHES, APTA, NEASC, NLNAC[24]  
8  1784[25]         APTA, NEASC, NLNAC[25]  
9  1908[26]                      NEASC[26]

### Removing unwanted data

In [40]:
data=df.drop(['Accreditation[16]'], axis=1)
data.rename(columns={"Location[note 1]": "Location", "Control[1]": "Control", "Type[1]": "Type", "Enrollment[16]": "Enrollment"}, inplace=True)
for i in range (data['Enrollment'].shape[0]):
    data['Enrollment'][i] = re.sub("[\(\[].*?[\)\]]", "", data['Enrollment'][i])
for i in range (data['Founded'].shape[0]):
    data['Founded'][i] = re.sub("[\(\[].*?[\)\]]", "", data['Founded'][i])

data['Enrollment'] = data['Enrollment'].str.replace(',','')
data['Founded'] = data['Founded'].str.replace(',','')
data['Founded'][92] = 1975
data = data.astype({"Enrollment":'int64', "Founded":'int64'})

print('Data shape:',data.shape)
print('The data consists of {} colleges in {} different cities'.format(data.shape[0],np.count_nonzero(data['Location'].unique())))
data.head(10)

Data shape: (104, 6)
The data consists of 104 colleges in 53 different cities


School          Location  \
0             American International College       Springfield   
1                            Amherst College           Amherst   
2                         Anna Maria College            Paxton   
3                         Assumption College         Worcester   
4                             Babson College         Wellesley   
5               Bard College at Simon's Rock  Great Barrington   
6                        Bay Path University        Longmeadow   
7                          Bay State College            Boston   
8                             Becker College         Worcester   
9  Benjamin Franklin Institute of Technology            Boston   

                  Control                               Type  Enrollment  \
0  Private not-for-profit                Master's university        2177   
1  Private not-for-profit              Baccalaureate college        1817   
2  Private not-for-profit                Master's university        1455   
3  Private not-for-profit                Master's university        2813   
4  Private not-for-profit          Special-focus institution        3250   
5  Private not-for-profit  Baccalaureate/associate's college         354   
6  Private not-for-profit              Baccalaureate college        2370   
7              For-profit                Associate's college        1721   
8  Private not-for-profit              Baccalaureate college        1826   
9  Private not-for-profit          Special-focus institution         475   

   Founded  
0     1885  
1     1821  
2     1946  
3     1904  
4     1919  
5     1964  
6     1897  
7     1946  
8     1784  
9     1908

### Lets explore the data

In [41]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 6 columns):
School        104 non-null object
Location      104 non-null object
Control       104 non-null object
Type          104 non-null object
Enrollment    104 non-null int64
Founded       104 non-null int64
dtypes: int64(2), object(4)
memory usage: 5.0+ KB


First lets focus on the numeric features

In [42]:
data[['Enrollment', 'Founded']].describe()

Enrollment      Founded
count    104.000000   104.000000
mean    4780.480769  1913.009615
std     5934.864141    57.049713
min       18.000000  1636.000000
25%     1281.750000  1878.000000
50%     2496.000000  1918.000000
75%     6314.750000  1963.250000
max    32603.000000  1997.000000

In [46]:
print('The oldest college in Massachusetts is {} founded in {}'.format(data['School'][np.argmin(data['Founded'])],np.min(data['Founded'])))
print('\n')
print('The newest college in Massachusetts is {} founded in {}'.format(data['School'][np.argmax(data['Founded'])],np.max(data['Founded'])))
print('\n')
print('The college with the highest enrollment is {} with {} students'.format(data['School'][np.argmax(data['Enrollment'])],np.max(data['Enrollment'])))
print('\n')
print('The college with the lowest enrollment is {} with {} students'.format(data['School'][np.argmin(data['Enrollment'])],np.min(data['Enrollment'])))
print('\n')
print('The average enrollment for a college in Massachusetts is approximately {} students'.format(round(data['Enrollment'].mean())))
print('\n')
print('The total number of students enrolled in Massachusetts is {} students'.format(data['Enrollment'].sum()))

The oldest college in Massachusetts is Harvard University founded in 1636


The newest college in Massachusetts is Olin College founded in 1997


The college with the highest enrollment is Boston University with 32603 students


The college with the lowest enrollment is Conway School of Landscape Design with 18 students


The average enrollment for a college in Massachusetts is approximately 4780 students


The total number of students enrolled in Massachusetts is 497170 students


In [47]:
print('{} college was founded in the 17th century'.format(np.count_nonzero(data['Founded'].between(1601, 1700, inclusive = True))))
print('\n')
print('{} colleges were founded in the 18th century'.format(np.count_nonzero(data['Founded'].between(1701, 1800, inclusive = True))))
print('\n')
print('{} colleges were founded in the 19th century'.format(np.count_nonzero(data['Founded'].between(1801, 1900, inclusive = True))))
print('\n')
print('{} colleges were founded in the 20th century'.format(np.count_nonzero(data['Founded'].between(1901, 2000, inclusive = True))))
print('\n')

1 college was founded in the 17th century


2 colleges were founded in the 18th century


40 colleges were founded in the 19th century


61 colleges were founded in the 20th century




Now lets focus on the categorical features

In [48]:
data[['School', 'Location', 'Control', 'Type']].describe()

School Location                 Control  \
count                   104      104                     104   
unique                  104       53                       3   
top     Bay Path University   Boston  Private not-for-profit   
freq                      1       24                      72   

                             Type  
count                         104  
unique                          6  
top     Special-focus institution  
freq                           28

 - We can see that there are 104 distinct colleges in the dataset, in a total of 53 different locations
 - The location with most colleges is Boston with a total of 24 colleges.
 - Out of the 104 colleges, 72 are Private not-for-profit.
 - The most common type of college are Special-focus institutions.

In [49]:
data_location = data.groupby(['Location']).School.nunique()
data_location.sort_values(ascending=False).head()

Location
Boston         24
Worcester       8
Cambridge       6
Springfield     4
Wellesley       3
Name: School, dtype: int64

Most colleges are in Boston, Worcester and Cambridge

In [50]:
data_control = data.groupby(['Control']).School.nunique()
data_control.sort_values(ascending=False)

Control
Private not-for-profit    72
Public                    30
For-profit                 2
Name: School, dtype: int64

72 colleges are Private not-for-profit, 30 are Public and 2 are For-profit

In [51]:
data_type = data.groupby(['Type']).School.nunique()
data_type.sort_values(ascending=False)

Type
Special-focus institution            28
Baccalaureate college                21
Master's university                  20
Associate's college                  20
Research university                  14
Baccalaureate/associate's college     1
Name: School, dtype: int64

28 colleges are Special-focus institutions, 22 are Baccalaureate colleges, 20 are Master's universities, 20 are Associate's colleges and 14 are Research universities

### Lets prepare the data in a convenient way

In [52]:
data['Number of Colleges'] = data['Location'].map(data['Location'].value_counts())
data.head(10)

School          Location  \
0             American International College       Springfield   
1                            Amherst College           Amherst   
2                         Anna Maria College            Paxton   
3                         Assumption College         Worcester   
4                             Babson College         Wellesley   
5               Bard College at Simon's Rock  Great Barrington   
6                        Bay Path University        Longmeadow   
7                          Bay State College            Boston   
8                             Becker College         Worcester   
9  Benjamin Franklin Institute of Technology            Boston   

                  Control                               Type  Enrollment  \
0  Private not-for-profit                Master's university        2177   
1  Private not-for-profit              Baccalaureate college        1817   
2  Private not-for-profit                Master's university        1455   
3  Private not-for-profit                Master's university        2813   
4  Private not-for-profit          Special-focus institution        3250   
5  Private not-for-profit  Baccalaureate/associate's college         354   
6  Private not-for-profit              Baccalaureate college        2370   
7              For-profit                Associate's college        1721   
8  Private not-for-profit              Baccalaureate college        1826   
9  Private not-for-profit          Special-focus institution         475   

   Founded  Number of Colleges  
0     1885                   4  
1     1821                   3  
2     1946                   1  
3     1904                   8  
4     1919                   3  
5     1964                   1  
6     1897                   1  
7     1946                  24  
8     1784                   8  
9     1908                  24

In [53]:
result = data.groupby('Location', sort=True).agg( ','.join)
result = result.reset_index(drop=True)

result.head()

School  \
0  Amherst College,Hampshire College,University o...   
1                        Massachusetts School of Law   
2         Endicott College,Montserrat College of Art   
3  Bay State College,Benjamin Franklin Institute ...   
4                     Massachusetts Maritime Academy   

                                             Control  \
0  Private not-for-profit,Private not-for-profit,...   
1                             Private not-for-profit   
2      Private not-for-profit,Private not-for-profit   
3  For-profit,Private not-for-profit,Private not-...   
4                                             Public   

                                                Type  
0  Baccalaureate college,Baccalaureate college,Re...  
1                          Special-focus institution  
2      Master's university,Special-focus institution  
3  Associate's college,Special-focus institution,...  
4                              Baccalaureate college

### Lets just keep the cities with the names and number of colleges in each one

In [54]:
data_clean = pd.DataFrame(data[['Location','Number of Colleges']])
data_clean = data_clean.drop_duplicates()
data_clean.sort_values('Location',inplace=True)
data_clean.rename(columns={"Location": "City"}, inplace=True)
data_clean = data_clean.reset_index(drop=True)
data_clean['Colleges'] = result['School']

data_clean.head(10)

City  Number of Colleges  \
0      Amherst                   3   
1      Andover                   1   
2      Beverly                   2   
3       Boston                  24   
4       Bourne                   1   
5  Bridgewater                   1   
6     Brighton                   1   
7     Brockton                   1   
8    Brookline                   2   
9    Cambridge                   6   

                                            Colleges  
0  Amherst College,Hampshire College,University o...  
1                        Massachusetts School of Law  
2         Endicott College,Montserrat College of Art  
3  Bay State College,Benjamin Franklin Institute ...  
4                     Massachusetts Maritime Academy  
5                       Bridgewater State University  
6                              Saint John's Seminary  
7                        Massasoit Community College  
8  Boston Graduate School of Psychoanalysis,Helle...  
9  Cambridge College,Harvard University,Hult Inte...

### Now we find the coordinates of each city using geopy and append it to the dataframe

In [55]:
longitudes=[]
latitudes=[]

for i in data_clean['City']:
    address = i + ', Massachusetts'

    geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    latitudes.append(location.latitude)
    longitudes.append(location.longitude)

print('Coordinates ready')

Coordinates ready


In [56]:
data_clean['Latitude'] = latitudes
data_clean['Longitude'] = longitudes
print(data_clean.shape)
data_clean.head(10)

(53, 5)


City  Number of Colleges  \
0      Amherst                   3   
1      Andover                   1   
2      Beverly                   2   
3       Boston                  24   
4       Bourne                   1   
5  Bridgewater                   1   
6     Brighton                   1   
7     Brockton                   1   
8    Brookline                   2   
9    Cambridge                   6   

                                            Colleges   Latitude  Longitude  
0  Amherst College,Hampshire College,University o...  42.368566 -72.505714  
1                        Massachusetts School of Law  42.657170 -71.140878  
2         Endicott College,Montserrat College of Art  42.558428 -70.880049  
3  Bay State College,Benjamin Franklin Institute ...  42.360253 -71.058291  
4                     Massachusetts Maritime Academy  41.741217 -70.598920  
5                       Bridgewater State University  41.990379 -70.975043  
6                              Saint John's Seminary  42.350097 -71.156442  
7                        Massasoit Community College  42.083433 -71.018379  
8  Boston Graduate School of Psychoanalysis,Helle...  42.331764 -71.121163  
9  Cambridge College,Harvard University,Hult Inte...  42.375100 -71.105616

### Use geopy library to get the latitude and longitude values of the state of Massachusetts 

In [57]:
address = 'Massachusetts, United States'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Massachusetts, United States are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Massachusetts, United States are 42.3788774, -72.032366.


### Create a map of Massachusetts with the cities that have colleges superimposed on top

In [58]:
# create map of Massachusetts using latitude and longitude values
map_massachusetts = folium.Map(location=[latitude, longitude], zoom_start=9)

# add markers to map
for lat, lng, city in zip(data_clean['Latitude'], data_clean['Longitude'], data_clean['City']):
    label = '{}, {}'.format(city, 'Massachusetts')
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_massachusetts)  
    
map_massachusetts

### Define Foursquare Credentials and Version

In [59]:
# The code was removed by Watson Studio for sharing.

### Let's create a function to find the closest venues for all the cities

In [60]:
def getNearbyVenues(names, latitudes, longitudes, radius=5000):
    
    LIMIT=1000
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['City', 
                  'City Latitude', 
                  'City Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [61]:
#get all nearby venues
massachusetts_venues = getNearbyVenues(names=data_clean['City'],
                                 latitudes=data_clean['Latitude'],
                                 longitudes=data_clean['Longitude']
                                 )
print('Venues ready')

Venues ready


### Lets visualize the data extracted from Foursquare

In [62]:
print(massachusetts_venues.shape)
massachusetts_venues.head(10)

(4270, 7)


City  City Latitude  City Longitude                   Venue  \
0  Amherst      42.368566      -72.505714             Spirit Haus   
1  Amherst      42.368566      -72.505714          Amherst Cinema   
2  Amherst      42.368566      -72.505714          Amherst Coffee   
3  Amherst      42.368566      -72.505714         Antonio's Pizza   
4  Amherst      42.368566      -72.505714          Amherst Common   
5  Amherst      42.368566      -72.505714            Pita Pockets   
6  Amherst      42.368566      -72.505714               Lone Wolf   
7  Amherst      42.368566      -72.505714        Lord Jeffery Inn   
8  Amherst      42.368566      -72.505714           Amherst Books   
9  Amherst      42.368566      -72.505714  Emily Dickinson Museum   

   Venue Latitude  Venue Longitude    Venue Category  
0       42.373262       -72.502193      Liquor Store  
1       42.375219       -72.520377     Movie Theater  
2       42.375694       -72.520501       Coffee Shop  
3       42.376193       -72.519810       Pizza Place  
4       42.373922       -72.519379              Park  
5       42.377343       -72.519499  Halal Restaurant  
6       42.375735       -72.518509    Breakfast Spot  
7       42.374501       -72.518794             Hotel  
8       42.375742       -72.519568         Bookstore  
9       42.376222       -72.514377    History Museum

Let's check how many venues were returned for each city

In [63]:
massachusetts_venues.groupby('City').count().head()

City Latitude  City Longitude  Venue  Venue Latitude  \
City                                                            
Amherst            100             100    100             100   
Andover             95              95     95              95   
Beverly            100             100    100             100   
Boston             100             100    100             100   
Bourne              70              70     70              70   

         Venue Longitude  Venue Category  
City                                      
Amherst              100             100  
Andover               95              95  
Beverly              100             100  
Boston               100             100  
Bourne                70              70

#### Let's find out how many unique categories can be curated from all the returned venues

In [64]:
print('There are {} uniques categories.'.format(len(massachusetts_venues['Venue Category'].unique())))

There are 319 uniques categories.


### Now lets analyze each city

In [65]:
# one hot encoding
massachusetts_onehot = pd.get_dummies(massachusetts_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
massachusetts_onehot['City'] = massachusetts_venues['City'] 

# move neighborhood column to the first column
fixed_columns = [massachusetts_onehot.columns[-1]] + list(massachusetts_onehot.columns[:-1])
massachusetts_onehot = massachusetts_onehot[fixed_columns]

massachusetts_onehot.head()

City  ATM  Accessories Store  Afghan Restaurant  African Restaurant  \
0  Amherst    0                  0                  0                   0   
1  Amherst    0                  0                  0                   0   
2  Amherst    0                  0                  0                   0   
3  Amherst    0                  0                  0                   0   
4  Amherst    0                  0                  0                   0   

   Airport  American Restaurant  Animal Shelter  Antique Shop  Aquarium  \
0        0                    0               0             0         0   
1        0                    0               0             0         0   
2        0                    0               0             0         0   
3        0                    0               0             0         0   
4        0                    0               0             0         0   

   Arcade  Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0       0                 0            0           0                    0   
1       0                 0            0           0                    0   
2       0                 0            0           0                    0   
3       0                 0            0           0                    0   
4       0                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Automotive Shop  \
0                 0                   0                0                0   
1                 0                   0                0                0   
2                 0                   0                0                0   
3                 0                   0                0                0   
4                 0                   0                0                0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       0     0    0               0                 0   

   Beach  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0      0                0         0            0           0   
1      0                0         0            0           0   
2      0                0         0            0           0   
3      0                0         0            0           0   
4      0                0         0            0           0   

   Belgian Restaurant  Big Box Store  Bike Shop  Bistro  Boat or Ferry  \
0                   0              0          0       0              0   
1                   0              0          0       0              0   
2                   0              0          0       0              0   
3                   0              0          0       0              0   
4                   0              0          0       0              0   

   Bookstore  Boutique  Bowling Alley  Bowling Green  Boxing Gym  \
0          0         0              0              0           0   
1          0         0              0              0           0   
2          0         0              0              0           0   
3          0         0              0              0           0   
4          0         0              0              0           0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  Bridge  \
0                     0               0        0            0       0   
1                     0               0        0            0       0   
2                     0               0        0            0       0   
3                     0               0        0            0       0   
4                     0               0        0            0       0   

 

And let's examine the new dataframe size.

In [66]:
massachusetts_onehot.shape

(4270, 320)

#### Next, let's group rows by City and by taking the mean of the frequency of occurrence of each category

In [67]:
massachusetts_grouped = massachusetts_onehot.groupby('City').mean().reset_index()
massachusetts_grouped.head()

City  ATM  Accessories Store  Afghan Restaurant  African Restaurant  \
0  Amherst  0.0                0.0                0.0                 0.0   
1  Andover  0.0                0.0                0.0                 0.0   
2  Beverly  0.0                0.0                0.0                 0.0   
3   Boston  0.0                0.0                0.0                 0.0   
4   Bourne  0.0                0.0                0.0                 0.0   

   Airport  American Restaurant  Animal Shelter  Antique Shop  Aquarium  \
0      0.0             0.030000             0.0           0.0      0.00   
1      0.0             0.052632             0.0           0.0      0.00   
2      0.0             0.020000             0.0           0.0      0.00   
3      0.0             0.020000             0.0           0.0      0.01   
4      0.0             0.057143             0.0           0.0      0.00   

   Arcade  Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0     0.0               0.0         0.00    0.000000                  0.0   
1     0.0               0.0         0.00    0.010526                  0.0   
2     0.0               0.0         0.01    0.000000                  0.0   
3     0.0               0.0         0.00    0.000000                  0.0   
4     0.0               0.0         0.00    0.000000                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Automotive Shop  \
0          0.010000                0.00              0.0              0.0   
1          0.021053                0.00              0.0              0.0   
2          0.010000                0.00              0.0              0.0   
3          0.000000                0.01              0.0              0.0   
4          0.000000                0.00              0.0              0.0   

   BBQ Joint  Bagel Shop    Bakery      Bank       Bar  Baseball Field  \
0        0.0    0.010000  0.030000  0.010000  0.010000             0.0   
1        0.0    0.010526  0.021053  0.010526  0.010526             0.0   
2        0.0    0.000000  0.040000  0.000000  0.010000             0.0   
3        0.0    0.000000  0.080000  0.000000  0.000000             0.0   
4        0.0    0.000000  0.000000  0.014286  0.014286             0.0   

   Baseball Stadium     Beach  Bed & Breakfast  Beer Bar  Beer Garden  \
0          0.000000  0.000000              0.0       0.0         0.00   
1          0.000000  0.000000              0.0       0.0         0.00   
2          0.000000  0.020000              0.0       0.0         0.00   
3          0.000000  0.000000              0.0       0.0         0.01   
4          0.014286  0.028571              0.0       0.0         0.00   

   Beer Store  Belgian Restaurant  Big Box Store  Bike Shop  Bistro  \
0         0.0                0.00            0.0        0.0     0.0   
1         0.0                0.00            0.0        0.0     0.0   
2         0.0                0.01            0.0        0.0     0.0   
3         0.0                0.01            0.0        0.0     0.0   
4         0.0                0.00            0.0        0.0     0.0   

   Boat or Ferry  Bookstore  Boutique  Bowling Alley  Bowling Green  \
0       0.000000   0.020000       0.0       0.000000            0.0   
1       0.000000   0.010526       0.0       0.000000            0.0   
2       0.000000   0.000000       0.0       0.000000            0.0   
3       0.000000   0.000000       0.0       0.000000            0.0   
4       0.014286   0.000000       0.0       0.014286            0.0   

   Boxing Gym  Brazilian Restaurant  Breakfast Spot   Brewery  Bridal Shop  \
0         0.0                   0.0        0.020000  0.010000          0.0   
1         0.0                   0.0        0.021053  0.010526          0.0   
2         0.0                   0.0        0.020000  0.020000          0.0   
3         0.0                   0.0        0.010000  0.010000          0.0   
4         0.0                   0.0   

#### Let's confirm the new size

In [68]:
massachusetts_grouped.shape

(53, 320)

Let's write a function to sort the venues in descending order.

In [69]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each city.

In [70]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['City']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
massachusetts_venues_sorted = pd.DataFrame(columns=columns)
massachusetts_venues_sorted['City'] = massachusetts_grouped['City']

for ind in np.arange(massachusetts_grouped.shape[0]):
    massachusetts_venues_sorted.iloc[ind, 1:] = return_most_common_venues(massachusetts_grouped.iloc[ind, :], num_top_venues)

massachusetts_venues_sorted.head()

City 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Amherst           Coffee Shop         Grocery Store        Sandwich Place   
1  Andover           Coffee Shop   American Restaurant           Pizza Place   
2  Beverly           Coffee Shop    Italian Restaurant                  Park   
3   Boston                  Park                Bakery    Seafood Restaurant   
4   Bourne    Seafood Restaurant            Donut Shop     Convenience Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Hotel   American Restaurant                Bakery   
1    Italian Restaurant        Sandwich Place            Restaurant   
2           Pizza Place                Bakery        Ice Cream Shop   
3                   Gym           Coffee Shop                 Hotel   
4   American Restaurant                  Park        Sandwich Place   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Pizza Place      Department Store          Liquor Store   
1  Gym / Fitness Center  Fast Food Restaurant            Donut Shop   
2                   Pub        Sandwich Place   Indie Movie Theater   
3         Historic Site           Pizza Place        Sandwich Place   
4            Restaurant                 Beach           Gas Station   

    10th Most Common Venue  
0           Breakfast Spot  
1             Burger Joint  
2                  Brewery  
3  New American Restaurant  
4           Breakfast Spot

### Now lets cluster the cities with colleges using k-means clustering

Run *k*-means to cluster the cities into 4 clusters.

In [71]:
# set number of clusters
kclusters = 4

massachusetts_grouped_clustering = massachusetts_grouped.drop('City', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(massachusetts_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 0, 0, 1, 2, 0, 1, 0, 0], dtype=int32)

In [72]:
# add clustering labels
massachusetts_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

massachusetts_merged = data_clean

# merge massachusetts_grouped with massachusetts_data to add latitude/longitude for each city
massachusetts_merged = massachusetts_merged.join(massachusetts_venues_sorted.set_index('City'), on='City')

massachusetts_merged.head()

City  Number of Colleges  \
0  Amherst                   3   
1  Andover                   1   
2  Beverly                   2   
3   Boston                  24   
4   Bourne                   1   

                                            Colleges   Latitude  Longitude  \
0  Amherst College,Hampshire College,University o...  42.368566 -72.505714   
1                        Massachusetts School of Law  42.657170 -71.140878   
2         Endicott College,Montserrat College of Art  42.558428 -70.880049   
3  Bay State College,Benjamin Franklin Institute ...  42.360253 -71.058291   
4                     Massachusetts Maritime Academy  41.741217 -70.598920   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop         Grocery Store   
1               1           Coffee Shop   American Restaurant   
2               0           Coffee Shop    Italian Restaurant   
3               0                  Park                Bakery   
4               1    Seafood Restaurant            Donut Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Sandwich Place                 Hotel   American Restaurant   
1           Pizza Place    Italian Restaurant        Sandwich Place   
2                  Park           Pizza Place                Bakery   
3    Seafood Restaurant                   Gym           Coffee Shop   
4     Convenience Store   American Restaurant                  Park   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                Bakery           Pizza Place      Department Store   
1            Restaurant  Gym / Fitness Center  Fast Food Restaurant   
2        Ice Cream Shop                   Pub        Sandwich Place   
3                 Hotel         Historic Site           Pizza Place   
4        Sandwich Place            Restaurant                 Beach   

  9th Most Common Venue   10th Most Common Venue  
0          Liquor Store           Breakfast Spot  
1            Donut Shop             Burger Joint  
2   Indie Movie Theater                  Brewery  
3        Sandwich Place  New American Restaurant  
4           Gas Station           Breakfast Spot

### Finally, let's visualize the resulting clusters

In [73]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=9)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(massachusetts_merged['Latitude'], massachusetts_merged['Longitude'], massachusetts_merged['City'], massachusetts_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

#### Cluster 0

In [74]:
massachusetts_merged.loc[massachusetts_merged['Cluster Labels'] == 0, massachusetts_merged.columns[[0] + [1] + [2] + list(range(5, massachusetts_merged.shape[1]))]]

City  Number of Colleges  \
0           Amherst                   3   
2           Beverly                   2   
3            Boston                  24   
6          Brighton                   1   
8         Brookline                   2   
9         Cambridge                   6   
10    Chestnut Hill                   2   
19       Framingham                   1   
29          Medford                   1   
31          Needham                   1   
32           Newton                   2   
34    North Andover                   1   
40          Roxbury                   1   
41            Salem                   1   
45          Waltham                   2   
46        Wellesley                   3   
48  West Barnstable                   1   
50           Weston                   2   

                                             Colleges  Cluster Labels  \
0   Amherst College,Hampshire College,University o...               0   
2          Endicott College,Montserrat College of Art               0   
3   Bay State College,Benjamin Franklin Institute ...               0   
6                               Saint John's Seminary               0   
8   Boston Graduate School of Psychoanalysis,Helle...               0   
9   Cambridge College,Harvard University,Hult Inte...               0   
10                  Boston College,Pine Manor College               0   
19                        Framingham State University               0   
29                                   Tufts University               0   
31                                       Olin College               0   
32                      Hebrew College,Lasell College               0   
34                                  Merrimack College               0   
40                          Roxbury Community College               0   
41                             Salem State University               0   
45             Bentley University,Brandeis University               0   
46  Babson College,Massachusetts Bay Community Col...               0   
48                         Cape Cod Community College               0   
50  Pope St. John XXIII National Seminary,Regis Co...               0   

   1st Most Common Venue 2nd Most Common Venue    3rd Most Common Venue  \
0            Coffee Shop         Grocery Store           Sandwich Place   
2            Coffee Shop    Italian Restaurant                     Park   
3                   Park                Bakery       Seafood Restaurant   
6            Pizza Place                Bakery            Grocery Store   
8                   Park           Pizza Place                   Bakery   
9                 Bakery                  Café  New American Restaurant   
10        Ice Cream Shop           Pizza Place      American Restaurant   
19         Grocery Store                Bakery   Furniture / Home Store   
29                Bakery                  Café       Italian Restaurant   
31           Pizza Place                  Park       Italian Restaurant   
32        Ice Cream Shop   American Restaurant             Burger Joint   
34           Coffee Shop    Italian Restaurant              Pizza Place   
40    Italian Restaurant   American Restaurant                   Bakery   
41                  Park           Pizza Place              Coffee Shop   
45   American Restaurant    Italian Restaurant       Mexican Restaurant   
46           Coffee Shop   Japanese Restaurant       Italian Restaurant   
48                 Beach               Airport               Playground   
50                 Hotel           Coffee Shop               Steakhouse   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                  Hotel   American Restaurant                Bakery   
2            Pizza Place                Bakery        Ice Cream Shop   
3                    Gym           Coffee Shop                 Hotel   
6         Ice Cream Shop                 Trail                  Park   
8                  Trai

#### Cluster 1

In [75]:
massachusetts_merged.loc[massachusetts_merged['Cluster Labels'] == 1, massachusetts_merged.columns[[0] + [1] + [2] + list(range(5, massachusetts_merged.shape[1]))]]

City  Number of Colleges  \
1            Andover                   1   
4             Bourne                   1   
7           Brockton                   1   
11          Chicopee                   1   
13           Danvers                   1   
14         Dartmouth                   1   
17        Fall River                   1   
18         Fitchburg                   1   
20          Franklin                   1   
22  Great Barrington                   1   
23        Greenfield                   1   
24         Haverhill                   2   
25           Holyoke                   1   
26        Longmeadow                   1   
27            Lowell                   1   
28   Lowell, Bedford                   1   
30            Milton                   2   
33       North Adams                   1   
35       Northampton                   1   
38        Pittsfield                   1   
39            Quincy                   2   
42      South Hadley                   1   
43    South Hamilton                   1   
44       Springfield                   4   
47            Wenham                   1   
49         Westfield                   1   
51      Williamstown                   1   
52         Worcester                   8   

                                             Colleges  Cluster Labels  \
1                         Massachusetts School of Law               1   
4                      Massachusetts Maritime Academy               1   
7                         Massasoit Community College               1   
11                                       Elms College               1   
13                      North Shore Community College               1   
14              University of Massachusetts Dartmouth               1   
17                          Bristol Community College               1   
18                         Fitchburg State University               1   
20                                       Dean College               1   
22                       Bard College at Simon's Rock               1   
23                       Greenfield Community College               1   
24  Northern Essex Community College,Northpoint Bi...               1   
25                          Holyoke Community College               1   
26                                Bay Path University               1   
27                 University of Massachusetts Lowell               1   
28                        Middlesex Community College               1   
30               Boston Baptist College,Curry College               1   
33              Massachusetts College of Liberal Arts               1   
35                                      Smith College               1   
38                        Berkshire Community College               1   
39            Eastern Nazarene College,Quincy College               1   
42                              Mount Holyoke College               1   
43                Gordon–Conwell Theological Seminary               1   
44  American International College,Springfield Col...               1   
47                                     Gordon College               1   
49                         Westfield State University               1   
51                                   Williams College               1   
52  Assumption College,Becker College,Clark Univer...               1   

            1st Most Common Venue         2nd Most Common Venue  \
1                     Coffee Shop           American Restaurant   
4              Seafood Restaurant                    Donut Shop   
7                      Donut Shop             Convenience Store   
11                    Pizza Place           American Restaurant   
13            American Restaurant                Sandwich Place   
14                    Pizza Place                Clothing Store   
17                     Restaurant           American Restaurant   
18                    Coffee Shop             Convenience Store   
20            Amer

#### Cluster 2

In [76]:
massachusetts_merged.loc[massachusetts_merged['Cluster Labels'] == 2, massachusetts_merged.columns[[0] + [1] + [2] + list(range(5, massachusetts_merged.shape[1]))]]

City  Number of Colleges                           Colleges  \
5   Bridgewater                   1       Bridgewater State University   
15       Dudley                   1                    Nichols College   
16       Easton                   1                  Stonehill College   
21      Gardner                   1  Mount Wachusett Community College   
36       Norton                   1                    Wheaton College   
37       Paxton                   1                 Anna Maria College   

    Cluster Labels 1st Most Common Venue    2nd Most Common Venue  \
5                2            Donut Shop              Pizza Place   
15               2            Donut Shop              Pizza Place   
16               2         Grocery Store                 Pharmacy   
21               2            Donut Shop                 Pharmacy   
36               2            Donut Shop              Pizza Place   
37               2            Donut Shop  State / Provincial Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5         Sandwich Place              Pharmacy     Convenience Store   
15     Convenience Store            Restaurant        Sandwich Place   
16        Sandwich Place           Pizza Place            Donut Shop   
21           Gas Station        Sandwich Place        Discount Store   
36            Restaurant              Pharmacy           Video Store   
37           Pizza Place         Moving Target             Gastropub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5          Grocery Store           Coffee Shop       Thai Restaurant   
15                 Trail              Pharmacy    Seafood Restaurant   
16                   Gym    Salon / Barbershop   American Restaurant   
21            Restaurant           Pizza Place   American Restaurant   
36   American Restaurant  Gym / Fitness Center        Ice Cream Shop   
37            Campground           Golf Course                Market   

   9th Most Common Venue 10th Most Common Venue  
5      Mobile Phone Shop         Breakfast Spot  
15    Chinese Restaurant              Nightclub  
16           Golf Course           Liquor Store  
21   Fried Chicken Joint          Grocery Store  
36     Convenience Store         Sandwich Place  
37  Gym / Fitness Center         Breakfast Spot

#### Cluster 3

In [77]:
massachusetts_merged.loc[massachusetts_merged['Cluster Labels'] == 3, massachusetts_merged.columns[[0] + [1] + [2] + list(range(5, massachusetts_merged.shape[1]))]]

City  Number of Colleges                           Colleges  \
12  Conway                   1  Conway School of Landscape Design   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12               3                 River                  Farm   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12  Construction & Landscaping                 Trail    Photography Studio   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
12                   Bar     Food & Drink Shop                  Food   

   9th Most Common Venue 10th Most Common Venue  
12           Flower Shop            Flea Market